pkg/dockerregistry/server/repositorymiddleware.go

```diff
@@ -345,31 +255,47 @@ func (r *repository) Put(manifest *schema1.SignedManifest) error {
 				},
 			},
 			DockerImageReference: fmt.Sprintf("%s/%s/%s@%s", r.registryAddr, r.namespace, r.name, dgst.String()),
-			DockerImageManifest:  string(manifest.Raw),
+			DockerImageManifest:  string(payload),
 		},
 	}
 
-	if err := r.fillImageWithMetadata(manifest, &ism.Image); err != nil {
-		return err
+	for _, option := range options {
+		if opt, ok := option.(distribution.WithTagOption); ok {
+			ism.Tag = opt.Tag
+			break
+		}
+	}
+
+	if deserializedManifest, ok := manifest.(*schema2.DeserializedManifest); ok {
+		configBytes, err := r.Blobs(r.ctx).Get(r.ctx, deserializedManifest.Config.Digest)
+		if err != nil {
+			context.GetLogger(r.ctx).Errorf("Failed to get image config %s", deserializedManifest.Config.Digest.String())
+			return digest.Digest(""), err
+		}
+		ism.Image.DockerConfigImage = string(configBytes)
```

Comment:
> We have no control over it. He uploads to us as a blob. Client requests it in the same manner.

Hmm, my idea was to check media type passed in a descriptor to [`blobWriter.Commit()`](https://github.com/miminar/distribution/blob/respect-errors/registry/storage/blobwriter.go#L51), but I see that the descriptor accompanied with the config is just `application/octet-stream` which isn't a realiable indicator of what is being dealt with.